betaflight/betaflight
# Throttlebased EzLanding

An alternative, throttle based, method for reducing "bounciness". 
The original method as described by @ctzsnooze in #11998 uses motor output clipping.  
This PR also contains that method for easier comparison. 

The method implemented here tries to solve some issues with the motor clipping method:

* Further reduced bounciness when landing at just below hover throttle
* iTerm windup should be handled better
* Cross talk should be reduced since the outputs are scaled, rather than clipped
* Better interaction with dynamic idle, possibly preventing some of the instability on throttle chops

## Some definitions
* **effective throttle** the average motor output as a percentage of maximum motor output.
* **requested throttle** the throttle requested by the pilot.
* **PID motor outputs** the motor output values requested by the PID controllers.

## How the method works compared to airmode off and the previous PR
Bounciness during landings basically happens since large PID outputs from collisions causes the throttle to be adjusted upwards.
In the original PR the effective throttle is prevented from increasing beyond a limit by simply clipping the motor outputs if the throttle is low and the other sticks are close to centered.
The method in this PR instead limits how much the throttle can be adjusted upwards.

Unlike the old (pre 4.4) airmode off this method will scale the PID motor outputs, instead of clipping them if they do not fit.
Unlike the new 4.4 airmode off the PID motor outputs will only be scaled if they would otherwise be clipped.

**Algorithm outline** 
1. Calculate max allowable throttle increase according to some formula.
2. Calculate the "airmode throttle" the same way that it is currently done. 
3. If the airmode throttle is higher than the allowable throttle limit, set the throttle to that limit and scale the motor output from the PIDs so that the lowest value + the calculated throttle = minimum allowed motor output.
If the airmode throttle is below the limit, don't scale anything.
4. Make sure that the anti windup is aware of the reduced authority range.

*some formula* here is almost the same formula that is used in the original PR to calculate the limit motor output limit.
The one change is that throttle is not factored in directly.
Instead the maximum of the requested throttle and the output from the formula is used.
Since the formula have a minimum output at `ez_landing_limit`, the PIDs will have a bit of wiggle room even with requested throttle at 0 and stick centered. 

## Usage 
To activate the algorithm `mixer_type` should be set to `EZLANDING`

For testing: set `debug_mode` to `EZLANDING`

### Settings
* `ez_landing_limit`  _Default: 5, Range: 0-75_. the mixer will always be allowed to increase the throttle to at least this value in percent.
Increase if the quad feels unstable at zero throttle, decrease if the quad feels too bouncy with the RPY sticks centered. 
* `ez_landing_threshold` _Default: 25, Range: 0-200_.  Deflecting the roll, pitch or yaw stick this much in percent will effectively give the mixer full authority over the throttle (i.e. as whit airmode on with the Legacy mixer type). The authority of the mixer scales linearly when the max deflection is bellow this value. 

### Debug values
* **Field 0** Scaling of the motor range. If this is 10000 the throttle and motor output range is currently the same as it would be with the Legacy mixer and airmode on. If  the value is lower the motor output range is being scaled down to avoid increasing the throtte.
* **Field 1** Same as field 1 but if the throttle set point was 0.
* **Field 2** The maximum throttle that the mixer is allowed to request.
* **Field 3** Effective throttle, the throttle that the mixer is actually requesting.

### Old Usage
To activate the algorithm `mixer_type` should be set to `EZLANDING_THROTTLE`
For comparison the algorithm from the previous PR can be activated by setting `mixer_type` to `EZLANDING_CLIP`.

The default values for `ez_landing_limit` and `ez_landing_threshold` are the same as in the original PR (0.1 and 0.2).
The throttle based algorithm uses the values slightly differently so if `EZLANDING_THROTTLE` is set, they will be automatically adjusted to make the authority available to the PIDs as close as possible to the original PR (ez_landing_limit will be halved and ez_landing_threshold doubled when used in the code).

Setting `mixer_type` to `LEGACY` should turn of both modes and make the quad behave as normal.

## Testing instructions
See the original PR for basic testing instructions.

Pay special attention to: 
* instability at throttle chops and long dead drops
* Bounciness when landing at slightly below hover throttle

## Known issues
* Most likely not compatible with 3D quads in the current shape
* Sudden onset of PID authority. Set-point or set-point movement might be a better way to estimate how much priority the PIDs should be given 

Please mark this PR with *Testing Required* and *Don't Merge*
